Libs/libMacGitverCore/Widgets/ProgressDlg.cpp

```diff
@@ -0,0 +1,261 @@
+                            "x1: 0, y1: 0, x2: 0, y2: 1,"
+                            "stop: 0 %1,"
+                            "stop: 1 %2);"
+                            "}")
+                        .arg(barColor.lighter(150).name()).arg(barColor.name())
+                        );
+        }
+
+    public:
+        bool    mActive;
+        qreal   mPercentage;
+
+        Steps   mSteps;
+    };
+
+}
+
+
+ProgressDlg::ProgressDlg()
+    : BlueSky::Dialog()
+    , ui( new Ui::ProgressDlg )
+    , mDone( false )
+{
+    ui->setupUi( this );
+
+    QPushButton* close = ui->buttonBox->button(QDialogButtonBox::Close);
+    close->setEnabled( false );
+    connect(close, &QPushButton::clicked, this, &ProgressDlg::close);
+
+    connect(&mUpdater, &QTimer::timeout, this, &ProgressDlg::updateActivities);
+    mUpdater.start(250);
+}
+
+ProgressDlg::~ProgressDlg()
+{
+    delete ui;
+}
+
+int ProgressDlg::updateInterval() const
+{
+    return mUpdater.interval();
+}
+
+void ProgressDlg::setUpdateInterval(int msec)
+{
+    if (mUpdater.isActive()) {
+        // restart timer with the new interval
+        mUpdater.start(msec);
+    }
+    else {
+        mUpdater.setInterval(msec);
+    }
+}
+
+void ProgressDlg::addActivity(const QString& description, QObject* activity,
+                              const StepInfo::List& steps)
+{
+    Q_ASSERT(activity);
+
+    Private::ProgressWdgt* a = new Private::ProgressWdgt(description);
+    mActivities[activity] = a;
+
+    QTreeWidgetItem* activityItem(new QTreeWidgetItem);
+    ui->treeProgress->addTopLevelItem(activityItem);
+    ui->treeProgress->setItemWidget(activityItem, 0, a);
+
+    foreach (const StepInfo& sir, steps) {
+        Private::ProgressWdgt* s = new Private::ProgressWdgt(sir.desc);
+        a->mSteps[sir.name] = s;
+
+        QTreeWidgetItem* stepItem(new QTreeWidgetItem);
+        activityItem->addChild(stepItem);
+        ui->treeProgress->setItemWidget(stepItem, 0, s);
+    }
+
+    a->txtStatusInfo->setText(tr("Running..."));
+}
+
+void ProgressDlg::setStatusInfo(QObject* activity, const QString& step,
+                                const QString& text)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->txtStatusInfo->setText(text);
+}
+
+void ProgressDlg::setPercentage(QObject* activity, const QString& step,
+                                qreal percent)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->mPercentage = qMin( qMax(percent, 0.), 1. ) * 100.;
+}
+
+void ProgressDlg::closeEvent( QCloseEvent* ev )
+{
+    if( !mDone )
+    {
+        ev->ignore();
+        return;
+    }
+
+    QDialog::closeEvent( ev );
+}
+
+void ProgressDlg::finished(QObject* activity)
+{
+    Private::ProgressWdgt* a = mActivities[activity];
+    Q_ASSERT(a);
+
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Finished!"));
+    a->resultChanged(Private::ProgressWdgt::Result::Ok);
```

Comment:
Yes, this I actually want to use `enum class` from now on, and change the traditional way - thanks! :)